### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by PlutoniumHeart**
No more assertion failures in the 2D environment, (Some of them are ignored)
My test result: using FF6 installation, TacRef can load texture, FF5.5 Can't.

### CODE REVIEW

**Comment by PeterMeyer** [comment]
Hi @PlutoniumHeart 
Nice Job. Will fech your code soon and check this out.

File: src/graphics/bsplib/objectlod.cpp [hunk 1]
```diff
@@ -524,7 +530,9 @@ void ObjectLOD::Reference(void)
 void ObjectLOD::Release(void)
 {
     // Dereference the object, and eventually uload
-    refCount--;
+    if (refCount>0)
+        refCount--;
```
Comment (vinnydiehl) [inline src/graphics/bsplib/objectlod.cpp:534]: Hey, thanks for some of the help cleaning up the syntax in this commit. These lines should be:

``` C++
if (refCount > 0)
    --refcount;
```

File: src/graphics/bsplib/texbank.cpp [hunk 2]
```diff
@@ -262,14 +262,18 @@ void TextureBankClass::Reference(int id)
     // If we already have the data, just verify that fact. Otherwise, load it.
     if (isLoaded)
     {
-        ShiAssert(TexturePool[id].tex.imageData || TexturePool[id].tex.TexHandle() || deferredLoadState);
+        //ShiAssert(TexturePool[id].tex.imageData || TexturePool[id].tex.TexHandle() || deferredLoadState);
+        //if(!(TexturePool[id].tex.imageData || TexturePool[id].tex.TexHandle() || deferredLoadState))
+        //    TexturePool[id].refCount--;
         //LeaveCriticalSection(&ObjectLOD::cs_ObjectLOD);
```
Comment (vinnydiehl) [inline src/graphics/bsplib/texbank.cpp:268]: All of this commented out code should go before we merge this.

File: src/graphics/bsplib/texbank.cpp [hunk 3]
```diff
@@ -262,14 +262,18 @@ void TextureBankClass::Reference(int id)
...
     }
     else
     {
         ShiAssert(TexFileMap.IsReady());
         ShiAssert(CompressedBuffer);
-        ShiAssert(TexturePool[id].tex.imageData == NULL);
+        //ShiAssert(TexturePool[id].tex.imageData == NULL);
```
Comment (vinnydiehl) [inline src/graphics/bsplib/texbank.cpp:274]: Kill this commented out code too please.

File: src/graphics/ddstuff/imagebuf.cpp [hunk 4]
```diff
@@ -889,7 +889,9 @@ void ImageBuffer::Compose(ImageBuffer *srcBuffer, RECT *dstRect, RECT *srcRect)
         else hr = m_pBltTarget->Blt(srcRect, srcBuffer->m_pDDSBack, dstRect, DDBLT_WAIT, NULL);
     }
 
-    ShiAssert(SUCCEEDED(hr));
+    // Still not clear why it would fail 3 times with error 0x88760096 when entering Campaign
+    // or switching between Campaign scenarios.
+    //ShiAssert(SUCCEEDED(hr));
```
Comment (vinnydiehl) [inline src/graphics/ddstuff/imagebuf.cpp:894]: Once again, commented out code's gotta die.

File: src/graphics/texture/tex.cpp [hunk 5]
```diff
@@ -135,8 +135,12 @@ static int FindMsb(DWORD val)
     //InterlockedIncrement((long *)&m_dwNumHandles); // Number of instances
     //InterlockedExchangeAdd((long *)&m_dwTotalBytes,-sizeof(*this));
 #endif
+    if((texHandle != NULL) || (imageData != NULL))
+    {
+        FreeAll();
+    }
 
-    ShiAssert((texHandle == NULL) && (imageData == NULL));
+    //ShiAssert((texHandle == NULL) && (imageData == NULL));
```
Comment (vinnydiehl) [inline src/graphics/texture/tex.cpp:143]: Commented out code

File: src/graphics/texture/tex.cpp [hunk 6]
```diff
@@ -868,7 +872,8 @@ bool TextureHandle::Load(UInt16 mip, UInt chroma, UInt8 *TexBuffer, bool bDoNotL
             break;
 
         default:
-            ShiAssert(false);
+            ;
+            //ShiAssert(false);
```
Comment (vinnydiehl) [inline src/graphics/texture/tex.cpp:876]: Commented out code

File: src/graphics/texture/tex.cpp [hunk 7]
```diff
@@ -902,8 +907,8 @@ bool TextureHandle::Load(UInt16 mip, UInt chroma, UInt8 *TexBuffer, bool bDoNotL
             m_bImageDataOwned = true;
 
 #ifdef _DEBUG
-            InterlockedExchangeAdd((long *)&m_dwTotalBytes, dwSize);
-            InterlockedExchangeAdd((long *)&m_dwBitmapBytes, dwSize);
+            //InterlockedExchangeAdd((long *)&m_dwTotalBytes, dwSize);
+            //InterlockedExchangeAdd((long *)&m_dwBitmapBytes, dwSize);
```
Comment (vinnydiehl) [inline src/graphics/texture/tex.cpp:911]: Commented out code

File: src/sim/simdrive/simdrive.cpp [hunk 8]
```diff
@@ -959,14 +959,14 @@ void SimulationDriver::SetFrameDescription(int mSecPerFrame, int numMinorFrames)
 void SimulationDriver::WakeCampaignBase(int isUnit, CampBaseClass* baseEntity, TailInsertList *comps)
 {
     // pm
-#ifdef _DEBUG
-    int flightList = 0;
-#endif
+//#ifdef _DEBUG
+//    int flightList = 0;
+//#endif
```
Comment (vinnydiehl) [inline src/sim/simdrive/simdrive.cpp:964]: Commented out code

File: src/sim/simdrive/simdrive.cpp [hunk 9]
```diff
@@ -959,14 +959,14 @@ void SimulationDriver::SetFrameDescription(int mSecPerFrame, int numMinorFrames)
...
     SimBaseClass* theObject = NULL;
     int vehicles = 0, last_to_add = 0, woken = 0;
 
 
-    ShiAssert(flightList);
+    //ShiAssert(flightList);
```
Comment (vinnydiehl) [inline src/sim/simdrive/simdrive.cpp:969]: Commented out code

**Comment by vinnydiehl** [comment]
Thanks for the fixes, @PlutoniumHeart. I put line notes where ever there was commented out code, would you kindly remove all of it completely before we merge this?

**Comment by PlutoniumHeart** [comment]
@gbchaosmaster I am sorry about those comments, it's just for my convenience when I am not sure which part of take out, and then just forgot to delete them when the code is working. 

I just found a pretty big file I/O bug, which could be the one that's preventing us to enter the 3D environment, I will fix that too before my next commit.

**Comment by vinnydiehl** [comment]
Not a problem, as long as they're all out by the time I merge this, it's all good.

**Comment by vinnydiehl** [comment]
@PlutoniumHeart: I'm continuing work here in #17, and moved your commits to a branch in central so we can all chip in. Continue your work there instead of here on your fork (just rebase any commits that you may currently have on top of it).

**Pull request closed without merging**